### DISCUSSION TITLE
[5장] 표현식과 문 QUIZ

### DISCUSSION THREAD

**Discussion by junjeeong** [junjeeong/JS_Deepdive_book_study#7] [Q&A] [answered]
1. 준영: "+" 와 같은 연산 리터럴도 실제 메모리를 차지하는가?

2. 나름 : 개발자 도구에서 표현식이 아닌 문(변수 선언문, 조건문)을 실행하면 무엇이 출력되나요? 그것을 무엇이라고 부르는가??

3. 조한 : 자바스크립트에서 값의 평가가 이루어지는 시점은 언제인가?

5. 윤주 : 기능을 재사용하기 위해서 함수를 사용하는데 한번만 사용 가능한 '즉시실행함수'는 왜 사용하는가?

6. 준일 :  ";"를 사용해야 하는지 사용하지 말아야 하는지는 "이것"에 달려있는데 "이것"은 무엇인가요??

**Comment by junjeeong** [comment, marked as answer]
1. 자바스크립트 엔진은 코드를 파싱하고 해석하면서 리터럴과 토큰을 실제 메모리에 저장한다. 하지만 그 단위가 너무나 작기 때문에 메모리 사용량을 크게 증가시키지 않아 영향을 주지 않는다. 자바스크립트 엔진은 이처럼 메모리 사용량을 최적화 하고 성능을 향상시키기 위해서 작은 값들을 효율적으로 관리한다.

2. undefined라는 값이 출력된다. undefined는 실제로 문이 값을 반환하지 않는 경우의 기본적인 반환 값이기 때문이다.

3. 자바스크립트는 "지연 평가"라는 것을 지원함으로써 값이 정말로 필요로 할 때 비로소 평가가 이루어지며 이를 통해 코드를 최적화 하고 성능을 향상시킨다. 아래는 실제 코드 평가가 이루어지는 상황이다.
- 변수 할당: 변수에 값을 할당할 때 해당 값의 평가가 이루어집니다.
- 표현식 실행: 표현식이 실행될 때 그 안에 포함된 값들의 평가가 이루어집니다.
- 함수 호출: 함수를 호출할 때 인자로 전달되는 값들의 평가가 이루어집니다.
- 객체 속성 접근: 객체의 속성에 접근할 때 해당 속성의 값의 평가가 이루어집니다.
- 배열 요소 접근: 배열의 요소에 접근할 때 해당 요소의 값의 평가가 이루어집니다.

4. 즉시실행함수는 기존에 재사용성이 가능한 함수의 목적에서 벗어난다. 즉시실행함수는 한 번 실행되면 그 내부에 선언된 변수와 함수가 외부에서 접근할 수 없는 스코프 내에 존재하므로, 해당 함수를 다시 호출하거나 재사용하는 것은 불가능하다. 그럼에도 쓰는 이유는 기존에 전역 스코프를 오염시키지 않고 임시적인 스코프를 생성하여 변수들을 격리시키는 데 유용하게 사용하기 위함이다.

5. 회사 컨벤션,  세미콜론(;)을 사용해야 하는지 아니면 생략해도 되는지에 대한 여부는 회사나 프로젝트의 코딩 컨벤션에 따라 결정된다.

**Comment by ruukr8080** [comment]
- ### 준영:  "+" 와 같은 연산 리터럴도 실제 메모리를 차지하는가?


    "+" 와 같은 연산 리터럴은 실제 메모리를 차지하지 않는다.
    연산 리터럴은 컴파일 타임에 평가되어 값으로 변환되기 때문이다.** 


이해를 돕기 위한  코드.)

var a = 10 + 20;
var b = "+" + "=";

console.log(a); // 3
console.log(b); // +=



---------------------------------------------------------------------------------------------------------------------------------------------

- ### 나름 : 개발자 도구에서 표현식이 아닌 문(변수 선언문, 조건문)을 실행하면 무엇이 출력되나요? 그것을 무엇이라고 부르는가??



개발자 도구에서 표현식이 아닌 문(변수 선언문, 조건문)을 실행하면 아무것도 출력되지 않습니다. 
이러한 문은 값을 반환하지 않기 때문입니다. 



이해를 돕기 위한 코드.)


let a = 1;

if (a > 0) {
  // 아무것도 출력되지 않음
}

---------------------------------------------------------------------------------------------------------------------------------------------
- ### 조한 : 자바스크립트에서 값의 평가가 이루어지는 시점은 언제인가?


자바스크립트에서 값의 평가는 선언 시점과 사용 시점에 이루어집니다.

선언 시점: 변수나 상수가 선언되는 시점에 값이 평가됩니다.
사용 시점: 변수나 상수가 사용되는 시점에 값이 평가됩니다.

이해를 돕기 위한 코드.)

let a = 1;

let b = a + 1;

console.log(b); // 2


a는 선언 시점에 1로 평가되고, 변수 b는 사용 시점에 2로 평가됨

---------------------------------------------------------------------------------------------------------------------------------------------

- ###  윤주: 기능을 재사용하기 위해서 함수를 사용하는데 한번만 사용 가능한 '즉시실행함수'는 왜 사용하는가?.

- 콜백 함수로 사용

이벤트 핸들러, 애니메이션, 스크립트 로드 등에서 콜백 함수를 사용해야 할 때 즉시 실행 함수를 사용하면 코드가 간결해집니다.

- 함수의 매개변수로 사용
함수의 매개변수로 익명 함수를 전달해야 할 때 즉시 실행 함수를 사용하면 코드가 간결해집니다.
- 변수에 할당
변수에 익명 함수를 할당해야 할 때 즉시 실행 함수를 사용하면 코드가 간결해집니다.
- 익명 함수의 결과를 반환
익명 함수의 결과를 반환해야 할 때 즉시 실행 함수를 사용하면 코드가 간결해집니다.


즉시 실행 함수는 코드를 간결하게 할 수 있는 유용한 기능이지만, 한번만 사용할 수 있다는 단점이 있습니다. 따라서, 한번만 사용할 함수를 작성할 때 즉시 실행 함수를 사용하면 효과적입니다.



------------------------------------------------------------------------------------------------------------------------------------------------------
준일: ";"를 사용해야 하는지 사용하지 말아야 하는지는 코드의 가독성과 유지보수성에 달려 있습니다.

-코드의 가독성을 높이기 위해서 사용
코드의 가독성을 높이기 위해서는 코드를 한눈에 알아볼 수 있도록 해야 합니다. ";"를 사용하면 코드의 구분이 명확해져 코드의 가독성을 높일 수 있습니다.
-코드의 유지보수성을 높이기 위해서 사용
코드의 유지보수성을 높이기 위해서는 코드를 이해하기 쉽도록 해야 합니다. ";"를 사용하면 코드의 구분이 명확해져 코드의 유지보수성을 높일 수 있습니다.
